pkg/daemon/daemon.go

```diff
@@ -791,27 +791,60 @@ func (dn *Daemon) getNodeMachinePool() error {
 	return fmt.Errorf("getNodeMachinePool(): Failed to find the MCP of the node")
 }
 
+func (dn *Daemon) getDrainLock(ctx context.Context, done chan bool) {
+	var err error
+
+	lock := &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Name:      "config-daemon-draining-lock",
+			Namespace: namespace,
+		},
+		Client: dn.kubeClient.CoordinationV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: dn.name,
+		},
+	}
+
+	// start the leader election
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock:            lock,
+		ReleaseOnCancel: true,
+		LeaseDuration:   5 * time.Second,
+		RenewDeadline:   3 * time.Second,
+		RetryPeriod:     1 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				glog.V(2).Info("drainNode(): started leading")
+				for {
+					time.Sleep(1 * time.Second)
+					if dn.drainable {
+						glog.V(2).Info("drainNode(): no other node is draining")
+						err = dn.annotateNode(dn.name, annoDraining)
```

Comment:
using this mechanism, do we still need to annotate the node with `annotDraining` ?

Reply:
That is the trick. The leader election mechanism requires the leader to keep updating the Lease object. But in our case, the node may reboot itself, then lose leadership. So I use a 2 layers lock here. The node can only start draining with 2 conditions: 1) it becomes the leader 2) no other node is draining which is indicated by the annotation.

Reply:
@pliurh mind mentioning this in the commit message ? so its clear in the commit message how this mechanism is used to control node draining